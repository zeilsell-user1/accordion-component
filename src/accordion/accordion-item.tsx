import "./styles.css";

import { useState } from "react";

import type { Item, SubItem } from "./accordion.types";

interface AccordionItemProps {
  item: Item;
}

export const AccordionItem = ({ item }: AccordionItemProps): JSX.Element => {
  const [isVisible, setIsVisible] = useState(false);

  const renderSubMenu = (subItem: SubItem, index: number): JSX.Element => {
    if (!subItem?.enabled || !subItem?.url) {
      return (
        <div>
          <p>{index}</p>
        </div>
      );
    }

    if (subItem.description === undefined || subItem.description === null) {
      subItem.description = "";
    }

    return (
      <div key={subItem.key} className={"accordion-dropdown-container"}>
        <p className={"accordion-dropdown-text"}></p>
        <p className={"accordion-tropdown-text"}>
          <a href={subItem.url}>{subItem.title.substring(0, 15)}</a>
        </p>
      </div>
    );
  };

  const toggleVisible = (): void => {
    setIsVisible(!isVisible);
  };

  return (
    <div>
      {item.url ? (
        <div className="accordion-item-container">
          <p className={"accordion-dropdown-text"}>
            <a href={item.url}>{item.title.substring(0, 15)}</a>
          </p>
        </div>
      ) : (
        <div className="accordion-item-container">
          <p className={"accordion-dropdown-text"}>
            {item.title.substring(0, 15)}
          </p>
          <div className={"accordian-button"} onClick={toggleVisible}>
            {isVisible ? "▲" : "▼"}
          </div>
        </div>
      )}
      <div>
        {isVisible && item.subItems && (
          <div>
            {item.subItems.map((subItem, index) =>
              renderSubMenu(subItem, index)
            )}
          </div>
        )}
      </div>
    </div>
  );
};
