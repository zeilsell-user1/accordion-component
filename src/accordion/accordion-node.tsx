import "./styles.css";

import { useState } from "react";

import type { AccordionItem, AccordionSubItem } from "./accordion.types";

interface AccordionNodeProps {
  item: AccordionItem;
}

export const AccordionNode = ({ item }: AccordionNodeProps): JSX.Element => {
  const [isVisible, setIsVisible] = useState(false);

  const renderSubMenu = (subItem: AccordionSubItem): JSX.Element => {
    if (!subItem?.enabled || !subItem?.url) {
      return <div />;
    }

    if (subItem.description === undefined || subItem.description === null) {
      subItem.description = "";
    }

    return (
      <div key={subItem.key} className={"accordion-dropdown-container"}>
        <p className={"accordion-dropdown-text"}></p>
        <p className={"accordion-tropdown-text"}>
          <a href={subItem.url}>{subItem.title.substring(0, 15)}</a>
        </p>
      </div>
    );
  };

  const toggleVisible = (): void => {
    setIsVisible(!isVisible);
  };

  return (
    <div>
      <div className={"accordion-item-container"}>
        <p className={"accordion-dropdown-text"}>
          {item.url ? (
            <a href={item.url}>{item.title.substring(0, 15)}</a>
          ) : (
            <a>{item.title.substring(0, 15)}</a>
          )}
        </p>
        {(item?.subItems?.length != 0) && (
          <div className={"accordian-button"} onClick={toggleVisible}>
            {isVisible ? "▲" : "▼"}
          </div>
        )}
      </div>
      <div>
        {isVisible && item.subItems && (
          <div>{item.subItems.map((subItem) => renderSubMenu(subItem))}</div>
        )}
      </div>
    </div>
  );
};
