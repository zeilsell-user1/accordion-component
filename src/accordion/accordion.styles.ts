
import CSS from 'csstype';

export const AccordionContainer: CSS.Properties = {
  color: "blue",
  border: "1px solid black"
};

export const AccordionItemContainer: CSS.Properties = {
color: 'blue',
border: '1px solid black'
};

export const AccordionSubItemContainer: CSS.Properties = {
color: 'red',
border: '1px solid green',
display: 'block'
};

export const AccordianButton: CSS.Properties = {
color: 'white',
background: 'dodgerblue',
border: 'none',
padding: '12px 16px',
fontSize: '16px',
cursor: 'pointer'
};