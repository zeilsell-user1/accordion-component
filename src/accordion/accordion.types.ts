type SubItem = {
    enabled: boolean;
    title: string;
    description?: string;
    url: string;
  };
  
  type Item = {
    enabled: boolean;
    title: string;
    description?: string;
    subItems: SubItem[];
  };
  
  export { SubItem, Item };