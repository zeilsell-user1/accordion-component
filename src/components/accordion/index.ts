export { Accordion } from './accordion';
export { SubItem } from './accordion.types';