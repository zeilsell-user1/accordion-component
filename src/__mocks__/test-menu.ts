
import { AccordionItem, AccordionSubItem } from "../accordion";

const subItem1: AccordionSubItem = {
    key: "subitem1",
    enabled: true,
    title: "First option",
    description: "this is subitem description 1",
    url: "https://www.bbc.co.uk",
  };
  
  const subItem2: AccordionSubItem = {
    key: "subitem2",
    enabled: true,
    title: "Second option",
    description: "this is subitem description 2",
    url: "https://www.bbc.co.uk/sport",
  };
  
  const subItem3: AccordionSubItem = {
    key: "subitem3",
    enabled: true,
    title: "Third option",
    description: "this is subitem description 3",
    url: "https://www.bbc.co.uk/news",
  };
  
  const subItem4: AccordionSubItem = {
    key: "subitem4",
    enabled: true,
    title: "123456789012345",
    description: "123456789012345678901234567890",
    url: "https://www.bbc.co.uk",
  };
  
  const subItem5: AccordionSubItem = {
    key: "subitem5",
    enabled: true,
    title: "Fifth option",
    description: "this is subitem description 5",
    url: "https://www.bbc.co.uk/sport",
  };
  
  const subItem6: AccordionSubItem = {
    key: "subitem6",
    enabled: true,
    title: "Sixth option",
    url: "https://www.bbc.co.uk/news",
  };
  
  const subItemList1: AccordionSubItem[] = [subItem1, subItem2, subItem3];
  const subItemList2: AccordionSubItem[] = [subItem4, subItem5, subItem6];
  
  const item1: AccordionItem = {
    key: "item1",
    enabled: true,
    title: "123456789012345",
    description: "123456789012345678901234567890",
    subItems: subItemList1,
  };
  
  const item2: AccordionItem = {
    key: "item2",
    enabled: true,
    title: "Second Top",
    subItems: subItemList2,
    url: "https://www.bbc.co.uk/news",
  };
  
  const item3: AccordionItem = {
    key: "item3",
    enabled: true,
    title: "Third Top",
    description: "not cool stuff",
    url: "https://www.bbc.co.uk/news",
  };

export const itemList: AccordionItem[] = [item1, item2, item3];
