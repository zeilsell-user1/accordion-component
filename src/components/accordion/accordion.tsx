import React from "react";
import { SubItem } from "./accordion.types";

type Props = {
  title:string;
  description?:string;
  subItems:SubItem[];
  openIcon?:string;
  closeIcon?:string;
};

export const Accordion = ({
  title,
  description,
  subItems,
  openIcon,
  closeIcon,
  }: Props): JSX.Element => {

  return (
      <>
      </>
  )
}

