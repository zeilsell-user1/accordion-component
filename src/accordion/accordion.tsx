import "./styles.css";

import type { AccordionItem } from "./accordion.types";
import { AccordionNode } from "./accordion-node";

interface Props {
  items: AccordionItem[];
  background?: string;
  color?: string;
  backgroundDesc?: string;
  colorDesc?: string;
}

export const Accordion = ({
  items,
  background,
  color,
  backgroundDesc,
  colorDesc,
}: Props): JSX.Element => {
  const renderItem = (item: AccordionItem): JSX.Element => {
    if (!item?.enabled) {
      return <div key={item.key} />;
    }

    return (
      <div key={item.key}>
        <AccordionNode
          item={item}
          backgroundDesc={backgroundDesc}
          colorDesc={colorDesc}
        />
      </div>
    );
  };

  return (
    <div
      className="accordion-container"
      style={{ background: background, color: color }}
    >
      {items ? items.map((item) => renderItem(item)) : <div />}
    </div>
  );
};
