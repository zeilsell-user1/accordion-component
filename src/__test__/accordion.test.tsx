import { render } from "@testing-library/react";

import { itemList } from "../__mocks__/test-menu";
import { Accordion } from "../accordion";

describe("Accordion component", () => {
  it("should render", () => {
    const container = render(
      <Accordion
        items={itemList}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
