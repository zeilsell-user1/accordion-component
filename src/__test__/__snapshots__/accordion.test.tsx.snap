// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion component should render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="accordion-container"
      >
        <div>
          <div>
            <div
              class="accordion-item-container"
            >
              <p
                class="accordion-dropdown-text"
              >
                123456789012345
              </p>
              <p
                class="accordion-dropdown-text"
              >
                123456789012345678901234567890
              </p>
              <div
                class="accordian-button"
              >
                ▼
              </div>
            </div>
            <div />
          </div>
        </div>
        <div>
          <div>
            <div
              class="accordion-item-container"
            >
              <p
                class="accordion-dropdown-text"
              >
                Second Top
              </p>
              <p
                class="accordion-dropdown-text"
              />
              <div
                class="accordian-button"
              >
                ▼
              </div>
            </div>
            <div />
          </div>
        </div>
        <div>
          <div>
            <div
              class="accordion-item-container"
            >
              <p
                class="accordion-dropdown-text"
              >
                <a
                  href="https://www.bbc.co.uk/news"
                >
                  Third Top
                </a>
              </p>
              <p
                class="accordion-dropdown-text"
              >
                not cool stuff
              </p>
            </div>
            <div />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="accordion-container"
    >
      <div>
        <div>
          <div
            class="accordion-item-container"
          >
            <p
              class="accordion-dropdown-text"
            >
              123456789012345
            </p>
            <p
              class="accordion-dropdown-text"
            >
              123456789012345678901234567890
            </p>
            <div
              class="accordian-button"
            >
              ▼
            </div>
          </div>
          <div />
        </div>
      </div>
      <div>
        <div>
          <div
            class="accordion-item-container"
          >
            <p
              class="accordion-dropdown-text"
            >
              Second Top
            </p>
            <p
              class="accordion-dropdown-text"
            />
            <div
              class="accordian-button"
            >
              ▼
            </div>
          </div>
          <div />
        </div>
      </div>
      <div>
        <div>
          <div
            class="accordion-item-container"
          >
            <p
              class="accordion-dropdown-text"
            >
              <a
                href="https://www.bbc.co.uk/news"
              >
                Third Top
              </a>
            </p>
            <p
              class="accordion-dropdown-text"
            >
              not cool stuff
            </p>
          </div>
          <div />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
